Add  download-results.py and refactor download-data.py for modularity

jashapiro opened the pull request:
Here I am adding the ability to download results data from the workflow output bucket to the download script.

Closes #378
Closes #420

This turned into a bigger reorganization, as the previous version of this script was not very modular. So now there are a bunch of functions to do various parts of the process. The `main()` function is mostly tasked with parsing arguments and delegating to the other functions.

The overall approach here is that you can download the results of a specific module by specifying the `--module` or `--module-results` argument and then naming a module or modules in a comma separated list. These results can be refined by project or sample, but not by other filters, as I didn't want to make assumptions about file names. You may want to look at how I am doing the searches for those identifiers. They may need refinement, but they do work for the test case I have for now...

Results files are downloaded to `data/<release>/results/<module>` and organized however they were organized in the results bucket. But note that the bucket itself organized without the "results" subdirectory, as that would be redundant. However, I figured we could use the same bucket for the test data and the simulated results, so I gave used the prefix `test-results` there, to contrast separate the modules from the simulated data that is stored in `test`.

You can also list the available results modules with the `--list-results` or `--list-modules` command. (why both? because I kept messing up when testing).

For dry runs I added a bit more output to allow for checking the size of the download. Or at least the total size of the data set, because I didn't want to get into parsing the full output to get the exact list of files to be downloaded and then try to match that up with the `ls` output (which is the only place with the sizes). The main purpose of this is so that people have some idea of the size of the data they are requesting, as some (the merge data, I mean) are big!

Finally, I changed the `--include` option that I never liked to `--process-stage` which is a bit odd for bulk data, but I think it works better overall by not being quite so generic. Our online docs do not directly cover this option, so the only places that needed updating for that change were a couple of readmes.

We _will_ need to document the `--module-results` flag though, but I thought that would require a full separate issue (which I will file shortly!)

Questions for review:
- Does the download structure make sense or do you have any other suggestions on that front?
  - note that I have not actually tested the test data version, as we don't have any results there yet!
- Please do test with `--dryrun` (and without) to see if things are working as expected. Note that the only  projects in the `merge-sce` module at the moment are `SCPCS000007` and `SCPCS000012` and the latter is much smaller than the former!
- do all the arguments make sense in how they work together?
- please let me know if you run into any bugs or think error messages should be cleaned up!
  - When I made things modular, I went back to raising exceptions a few places, which makes things a bit uglier that just exiting. I could catch some of the exceptions, and make things a bit prettier or go back to just exiting...

jaclyn-taroni commented:
@jashapiro - what do you mean by this? There are two buckets for test data set up already.

> However, I figured we could use the same bucket for the test data and the simulated results, so I gave used the prefix `test-results` there, to contrast separate the modules from the simulated data that is stored in `test`.

jashapiro commented:
> @jashapiro - what do you mean by this? There are two buckets for test data set up already.
> 
> > However, I figured we could use the same bucket for the test data and the simulated results, so I gave used the prefix `test-results` there, to contrast separate the modules from the simulated data that is stored in `test`.

Yes, I guess there are now. I hadn't looked at how the `*-public-data` were set up, and was only considering the `openscpca--sim-data` bucket that I had been working with in the workflow to this point and its mirror `openscpca-temp-simdata`. 

My thought here was we can probably get away with just the one public bucket, and it might simplify management a bit. But either way is fine, and if the buckets already exist, we can use them!

allyhawkins commented on download-data.py:
```diff
@@ -54,35 +367,18 @@ def main() -> None:
+        default="",
+        help="The results modules to download."
+        " A comma separated list of results modules to download."
+        "\nWhen this option is selected, the `--process-stage` and `--include-reports` options are ignored.",
```
My understanding is that if these are used then you only download the results and not release, so I would specify that in this help message.

allyhawkins commented on download-data.py:
```diff
@@ -155,151 +477,124 @@ def main() -> None:
-        current_symlink.symlink_to(release)
-        print(f"Updated 'current' symlink to point to '{release}'.")
+    ### Download the data ###
+    if not args.module_results:
```
Looking at this I really do want a flag that specifies download type... I think the default could be release data, but I like being explicit about what type of data you are trying to get. Some people could think that by having a `--module_results` argument, you are adding download of results to download of a release, but these are separate. 

Another kind of wild thought, but what if we had two scripts... We can pull out the functions you have into a separate `utils` script that gets read in to each of the scripts. Then you have one that's just for downloading results from the workflow and one just for downloading release data.

allyhawkins commented on download-data.py:
```diff
@@ -4,20 +4,150 @@
+    if ls_result.returncode:
+        print(
+            "Error listing release versions from the OpenScPCA release bucket.",
+            " Ensure you have the correct AWS permissions to access OpenScPCA data.",
```
Just because I've seen this error message myself, can we update it to remind people to login or use the correct profile since that might be why they don't have access 😬

allyhawkins commented on download-data.py:
```diff
@@ -26,12 +156,189 @@ def add_parent_dirs(patterns: List[str], dirs: List[str]) -> List[str]:
+    update_current: bool = True,
+) -> None:
+    """
+    Download results for a specific release of OpenScPCA.
```
```suggestion
    Download workflow results for a specific release of OpenScPCA.
```

allyhawkins commented on download-data.py:
```diff
@@ -4,20 +4,150 @@
+    """
+    Convert a size in bytes to something human readable.
+    """
+    for unit in ["B", "KiB", "MiB", "GiB", "TiB"]:
+        if size < 1024.0 or unit == "TiB":
+            break
+        size /= 1024.0
+    if unit == "B":
+        return f"{size:.0f} B"
+    return f"{size:.2f} {unit}"
```
I don't think this is working as you expect? When I try to download the `merge-sce` results module with dry run I see a total size of 0 B, but the files are being listed.

allyhawkins replied:
Update, if I include a `--projects` argument then the size is correct. So this doesn't work if you want to download results for an entire module.

jashapiro replied:
Hmm...I am presumably doing something wrong with the globs, since I am trying to match the behavior of `aws`. Will look at this case!

jashapiro replied:
I believe I fixed this in f1442f3. If you want to test again, that would be great! Thank you for the detailed report, it made it easy to find my error!

allyhawkins replied:
Yes that fix works! I can now run a dry run for an entire module or a specific project within a module and the appropriate files are listed and the size is listed and appears to be correct. 🎉

jaclyn-taroni commented on download-data.py:
```diff
@@ -54,35 +371,18 @@ def main() -> None:
     )
     parser.add_argument(
-        "--format",
+        "--module-results",
```
Is this argument intended to work with `--format`? Because it doesn't appear to be respected in my testing, and the output of `--help` lead me to believe it would be supported.

jashapiro replied:
No, it isn't, because I what format the results might take seemed too hard to predict. In this case we have both SCE and Anndata, but I don't expect it to be the norm, so I wasn't sure what to do to handle it at this stage. Maybe a future enhancement to allow more selectivity from results downloads?

jaclyn-taroni replied:
Then the help text should say something about:

> These results can be refined by project or sample, but not by other filters, as I didn't want to make assumptions about file names.

Rather than:

>  When this option is selected, the `--process-stage` and `--include-reports` options are ignored

jaclyn-taroni reviewed (commented):
I am returning an observation from testing, which, in combination with @allyhawkins comments elsewhere about the help message, does make me wonder if we should reconsider splitting the results download out to remove complexity? In weighing that question, maintainability and being able to explain this to users should both be considered.

jashapiro commented:
> I am returning an observation from testing, which, in combination with @allyhawkins comments elsewhere about the help message, does make me wonder if we should reconsider splitting the results download out to remove complexity? In weighing that question, maintainability and being able to explain this to users should both be considered.

I think splitting it up won't be too bad. Now that it is more modular, I can either pull out functions to a separate file or even more simply just import one script from the other to get the functions I need. That's the magic of `if __name__ == "__main__":`!

jashapiro commented:
Okay, I have split this up into two separate scripts: `download-data.py` and `download-results.py` where the results script imports from the data script to get some of its functionality.

I also updated the AWS error message to tell people what is probably wrong (profiles and login), and just output the AWS error itself rather than the stack trace. Some other help text was also updated. 

This is probably worth testing one more time before it goes in.

allyhawkins commented on .github/workflows/docker-image.yml:
```diff
@@ -1,9 +1,9 @@
-  push:
-    paths:
-      - 'Dockerfile'
+  # push:
+  #   paths:
+  #     - 'Dockerfile'
```
Just checking did you mean to comment this out?

jashapiro replied:
I did for now. It was trying to run in my branch, so we will need to modify it for proper automation later.

allyhawkins commented on download-results.py:
```diff
@@ -0,0 +1,318 @@
+    ls_result = subprocess.run(ls_cmd, capture_output=True, text=True)
+    if ls_result.returncode:  # authentication errors, usually
+        print(
+            "Error listing release versions from the OpenScPCA bucket.\n\n"
```
```suggestion
            "Error listing modules available in OpenScPCA workflow results bucket.\n\n"
```
I think this should be updated to reflect that you are listing modules?

jaclyn-taroni commented on download-results.py:
```diff
@@ -0,0 +1,318 @@
+    ls_cmd = ["aws", "s3", "ls", f"s3://{bucket}/{release}/"]
+    if profile:
+        ls_cmd += ["--profile", profile]
+    if bucket == TEST_RESULTS_BUCKET:
+        ls_cmd += ["--no-sign-request"]
```
Reporting a result I was not expecting (have not looked into it further yet):
```
% ./download-results.py --test-data

Error listing release versions from the OpenScPCA bucket.

Make sure you have the correct profile active (or use the --profile option) and run `aws sso login` before running this script.

AWS Error:  
Unable to locate credentials. You can configure credentials by running "aws configure".
```

allyhawkins replied:
When I run this, I get the following which makes me think it might have something to do with the ability to access the test bucket? 

```
./download-results.py --test-data --dryrun --profile openscpca
Error listing release versions from the OpenScPCA bucket.

Make sure you have the correct profile active (or use the --profile option) and run `aws sso login` before running this script.

AWS Error:  
An error occurred (AccessDenied) when calling the ListObjectsV2 operation: Access Denied
```

jaclyn-taroni replied:
This works just fine: `aws s3 ls s3://openscpca-test-workflow-results-public-access/test --no-sign-request`

`aws s3 ls s3://openscpca-test-workflow-results-public-access --profile openscpca` returns: 

```
An error occurred (AccessDenied) when calling the ListObjectsV2 operation: Access Denied
```

Which is what I would expect.

allyhawkins replied:
Confirming that this now works as expected with the updates.

allyhawkins reviewed (approved):
I did some testing of both scripts and everything looks good! I just had one comment about an error message that needs to be updated, but I don't need to see this again.

allyhawkins reviewed (approved):
🚀
